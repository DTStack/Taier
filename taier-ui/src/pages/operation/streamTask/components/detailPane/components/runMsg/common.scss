.vertex-diagram {
    background: #fff;
    width: 280px;
    user-select: none;

    .vertex-title {
        width: 100%;
        display: grid;
        grid-template-columns: auto 32px;
        align-items: center;
        height: 36px;
        line-height: 36px;
        border-radius: 4px 4px 0 0;
        color: #fff;
        padding: 0 0 0 12px;

        .t_title {
            overflow: hidden;
            text-overflow: ellipsis;
            white-space: nowrap;
            float: left;
        }
    }

    .vertex-content {
        padding-right: 36px;
        font-size: 12px;
        font-family: PingFangSC-Regular, "PingFang SC";
        font-weight: 400;
        color: rgba(102, 102, 102, 1);
        height: 198px;

        .tcolumn {
            overflow: hidden;
            text-overflow: ellipsis;
            display: -webkit-box;
            -webkit-line-clamp: 2;
            -webkit-box-orient: vertical;
            padding: 12px 15px 0;
            margin-bottom: 8px;
            text-align: left;
        }

        .t-text-col {
            height: 30px;
            line-height: 30px;

            .t-text-col-key {
                float: left;
                text-indent: 12px;

                .t-text-col_img {
                    width: 12px;
                    height: 12px;
                    margin: -2px 8px 0 0;
                }
            }

            .t-text-col-value {
                float: right;
                font-family: PingFangSC-Medium, "PingFang SC";
                font-weight: 500;
                color: rgba(51, 51, 51, 1);
            }
        }
    }
}
