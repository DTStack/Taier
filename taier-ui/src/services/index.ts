import { container } from 'tsyringe';
import EditorActionBarService from './editorActionBarService';
import ExecuteService from './executeService';
import CatalogueService from './catalogueService';
import BreadcrumbService from './breadcrumbService';
import RightBarService from './rightBarService';
import TaskRenderService from './taskRenderService';
import DataSourceService from './dataSourceService';
import TaskParamsService from './taskParamsService';

const editorActionBarService = container.resolve(EditorActionBarService);
const executeService = container.resolve(ExecuteService);
const catalogueService = container.resolve(CatalogueService);
const breadcrumbService = container.resolve(BreadcrumbService);
const rightBarService = container.resolve(RightBarService);
const taskRenderService = container.resolve(TaskRenderService);
const dataSourceService = container.resolve(DataSourceService);
const taskParamsService = container.resolve(TaskParamsService);

export {
    editorActionBarService,
    catalogueService,
    executeService,
    breadcrumbService,
    rightBarService,
    taskRenderService,
    dataSourceService,
    taskParamsService,
};
