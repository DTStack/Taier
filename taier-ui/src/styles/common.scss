/* Some styles used for multiply components */
.split-text {
    float: left;
    padding: 0 5px;
}

.task_offline_message {
    margin-bottom: 10px;
    padding: 10px;
    color: #666;
    font-size: 12px;
    background-color: #eef6fe;
}

.no-padding-modal {
    .ant-modal-body {
        padding: 0;
        font-size: 12px;
        line-height: 1.5;
    }
}

.ope-statistics {
    display: flex;
    flex: none;
    flex-wrap: nowrap;
    overflow-x: scroll;
    font-weight: bold;

    span {
        height: 32px;
        margin: 0 8px;
        font-size: 12px;
        line-height: 23px;
        white-space: nowrap;
    }
}

.status_overview_basic_block {
    display: inline-block;
    padding: 4px 20px;
    font-weight: 400;
    border-radius: 30px;
}

.status_overview_fail_font {
    @extend .status_overview_basic_block;
    color: var(--badge-failed-foreground);
    background: var(--badge-failed-background);
}

.status_overview_running_font {
    @extend .status_overview_basic_block;
    color: var(--badge-running-foreground);
    background: var(--badge-running-background);
}

.status_overview_grey_font {
    @extend .status_overview_basic_block;
    color: var(--badge-cancel-foreground);
    background: var(--badge-cancel-background);
}

.status_overview_yellow_font {
    @extend .status_overview_basic_block;
    color: var(--badge-pending-foreground);
    background: var(--badge-pending-background);
}

.status_overview_finished_font {
    @extend .status_overview_basic_block;
    color: var(--badge-finished-foreground);
    background: var(--badge-finished-background);
}

.status_overview_count_font {
    @extend .status_overview_basic_block;
    color: var(--badge-common-foreground);
    background: var(--badge-common-background);
}

.dt-refresh.ant-btn {
    width: 32px;
    height: 32px;
    padding: 0;
    color: var(--problemsInfoIcon-foreground);
    line-height: 32px;
    background: var(--editor-background);
    border: none;
    border-radius: 0;
    box-shadow: 0 2px 6px 0 var(--minimapSlider-hoverBackground);
}

.modal-body-nopadding {
    .ant-modal-body {
        padding: 0;
    }
}

// 全链路图
.modal-body--height100 {
    .ant-modal-content {
        height: 100%;

        .ant-modal-body {
            height: 528px;
        }
    }
}

.ellipsis {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
}
