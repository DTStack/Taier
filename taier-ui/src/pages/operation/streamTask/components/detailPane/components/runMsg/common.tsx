import { useState, useEffect, useMemo } from 'react';
import type { mxCell, mxGraph } from 'mxgraph';
import * as ReactDOMServer from 'react-dom/server';
import { Modal } from 'antd';
import MxGraphContainer from '@/components/mxGraph/container';
import { cloneDeep } from 'lodash';
import type { IFlinkJsonProps } from '.';
import classnames from 'classnames';
import './common.scss';

interface ICommonProps {
    flinkJson: IFlinkJsonProps[];
    loading: boolean;
    refresh: () => void;
}

type ISubJobVertices = IFlinkJsonProps['subJobVertices'][number];

interface ITopological {
    id: string;
    title: string;
    desc: string;
    delay: number;
    parallelism: number;
    received: number;
    sent: number;
    backPressured: number;
    subJobVertices?: ISubJobVertices[];
    childNode: ITopological[];
}

const getBackPressureColor = (backPressure: number, type: string) => {
    let color = '';
    if (backPressure >= 0 && backPressure <= 0.1) {
        // 正常
        if (type === 'title') {
            color = 'linear-gradient(298deg,rgba(90,203,255,1) 0%,rgba(36,145,247,1) 100%)';
        } else {
            color = '#F1FAFF';
        }
    } else if (backPressure >= 0.1 && backPressure <= 0.5) {
        // 低反压
        if (type === 'title') {
            color = 'linear-gradient(270deg,rgba(255,190,76,1) 0%,rgba(255,160,41,1) 100%)';
        } else {
            color = '#fff9f0';
        }
    } else if (backPressure > 0.5 && backPressure <= 1) {
        // 高反压
        if (type === 'title') {
            color = 'linear-gradient(270deg,#FF5F5C 0%,#ea8785 100%)';
        } else {
            color = '#ff5f5c4d';
        }
    }
    return color;
};

export default function Common({ flinkJson, loading, refresh }: ICommonProps) {
    const [graphData, setGraphData] = useState<ITopological[] | null>(null);
    // 当前选中的 cell
    const [selectedCell, setSelectedCell] = useState<mxCell | null>(null);
    const [visible, setVisible] = useState(false);

    const handleClick = (cell: mxCell, graph: mxGraph, event: React.MouseEvent<HTMLElement, MouseEvent>) => {
        if ((event.target as HTMLImageElement).nodeName === 'IMG') {
            if ((cell.value as ITopological)?.subJobVertices) {
                setVisible(true);

                setSelectedCell(cell);
            }
        }
    };

    const handleRenderCell = (cell: mxCell) => {
        if (cell.value) {
            const { backPressured, title, desc, delay, parallelism, received, sent, subJobVertices } =
                cell.value as ITopological;
            const res = ReactDOMServer.renderToString(
                <div className="vertex-diagram" style={{ background: getBackPressureColor(backPressured, 'title') }}>
                    <div className="vertex-title">
                        <div className="t_title">{title}</div>
                        {subJobVertices?.length && <img className="cursor-pointer" src="images/expand.svg" />}
                    </div>
                    <div
                        className="vertex-content"
                        style={{ background: getBackPressureColor(backPressured, 'content') }}
                    >
                        <div className="tcolumn" title={desc}>
                            {desc}
                        </div>
                        {[
                            { field: 'delay', title: 'Delay', data: `${delay}ms` },
                            { field: 'parallelism', title: 'Parallelism', data: parallelism },
                            { field: 'received', title: 'Record Received', data: received },
                            { field: 'sent', title: 'Record Sent', data: sent },
                            {
                                field: 'dashboard',
                                title: 'BackPressured(max)',
                                data: `${(backPressured * 100).toFixed(0)}%`,
                            },
                        ].map(({ field, title: fieldTitle, data }) => (
                            <div className="t-text-col" title={fieldTitle} key={field}>
                                <span className="t-text-col-key">
                                    <img src={`images/${field}.svg`} className="t-text-col_img" />${fieldTitle}
                                </span>
                                <span className="t-text-col-value">{data}</span>
                            </div>
                        ))}
                    </div>
                </div>
            );
            return res;
        }
        return '';
    };

    const convertArrToLinkedList = (arr?: ISubJobVertices[]): ITopological[] | null => {
        let res: ITopological[] | null = null;
        arr?.forEach(
            function (this: { point: ITopological[] | null }, vertice, index) {
                if (!this.point) {
                    res = [
                        {
                            id: vertice.id,
                            title: `Operator ${index}`,
                            desc: vertice.name,
                            delay: Object.values(vertice.delayMapList)[0],
                            parallelism: vertice.parallelism,
                            received: Object.values(vertice.recordsReceivedMap)[0],
                            sent: Object.values(vertice.recordsSentMap)[0],
                            // 子节点的最大反压数据从父节点获取
                            backPressured: selectedCell?.value?.backPressured || 0,
                            childNode: [],
                        },
                    ];
                    this.point = res[0].childNode;
                } else {
                    this.point[0] = {
                        id: vertice.id,
                        title: `Operator ${index}`,
                        desc: vertice.name,
                        delay: Object.values(vertice.delayMapList)[0],
                        parallelism: vertice.parallelism,
                        received: Object.values(vertice.recordsReceivedMap)[0],
                        sent: Object.values(vertice.recordsSentMap)[0],
                        backPressured: selectedCell?.value?.backPressured || 0,
                        childNode: [],
                    };
                    this.point = this.point[0].childNode;
                }
            },
            { point: res }
        );

        return res;
    };

    const handleRefreshSub = () => {
        setSelectedCell((c) => cloneDeep(c));
    };

    useEffect(() => {
        if (flinkJson.length) {
            setGraphData(
                flinkJson.map((json) => {
                    return {
                        id: json.jobVertexId,
                        title: `Chain ${json.subJobVertices.length} Operators`,
                        desc: json.jobVertexName,
                        delay: Number(Object.values(json.delayMap)[0]) || 0,
                        parallelism: json.parallelism,
                        received: json.recordsReceived,
                        sent: json.recordsSent,
                        backPressured: Math.max(...Object.values(json.backPressureMap)),
                        subJobVertices: json.subJobVertices,
                        childNode: [],
                    };
                })
            );
        }
    }, [flinkJson]);

    const subGraphData = useMemo(() => {
        if (selectedCell) {
            return convertArrToLinkedList((selectedCell.value as ITopological)?.subJobVertices);
        }
        return null;
    }, [selectedCell]);

    if (!graphData) {
        return (
            <div className={classnames('full-w', 'flex-1', 'flex', 'items-center', 'justify-center')}>
                暂未生成拓扑图
            </div>
        );
    }

    return (
        <>
            <MxGraphContainer<ITopological>
                config={{
                    tooltips: false,
                }}
                vertexSize={{
                    width: 280,
                    height: 234,
                }}
                onClick={handleClick}
                loading={loading}
                graphData={graphData}
                onRenderCell={handleRenderCell}
                onRefresh={refresh}
            />
            <Modal
                visible={visible}
                title="工作流"
                onCancel={() => setVisible(false)}
                footer={null}
                zIndex={1000}
                width={900}
                destroyOnClose
            >
                <div style={{ height: 500 }}>
                    <MxGraphContainer<ITopological>
                        config={{
                            tooltips: false,
                        }}
                        vertexSize={{
                            width: 280,
                            height: 234,
                        }}
                        direction="west"
                        onClick={handleClick}
                        loading={loading}
                        graphData={subGraphData}
                        onRenderCell={handleRenderCell}
                        onRefresh={handleRefreshSub}
                    />
                </div>
            </Modal>
        </>
    );
}
