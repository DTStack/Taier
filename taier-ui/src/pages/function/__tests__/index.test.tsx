import Function from '../';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import { Input } from 'antd';
import { FolderTree } from '@dtinsight/molecule/esm/workbench/sidebar/explore';
import { FUNCTOIN_ACTIONS } from '@/constant';
import api from '@/api';
import functionManagerService from '@/services/functionManagerService';
import { act } from 'react-dom/test-utils';
import '@testing-library/jest-dom';
import { TreeViewUtil } from '@dtinsight/molecule/esm/common/treeUtil';
import { fillFormContent } from './fixtures/utils';
import { catalogueService } from '@/services';

jest.useFakeTimers();
jest.mock('@/api');
jest.mock('@/services', () => ({
    catalogueService: {
        loadTreeNode: jest.fn(),
    },
}));
jest.mock('@/services/functionManagerService', () => ({
    setActive: jest.fn(),
    getState: jest.fn(() => ({
        folderTree: {
            data: [
                {
                    id: 1,
                },
            ],
        },
    })),
    get: jest.fn(() => ({
        data: {},
    })),
}));
jest.mock('@/services/resourceManagerService', () => ({
    checkNotDir: jest.fn().mockResolvedValue(true),
}));
jest.mock('@/components/folderPicker', () => ({ id, value, onChange }: any) => (
    <Input data-testid="mockFolderPicker" id={id} value={value} onChange={(e) => onChange(e.target.value)} />
));
// use the original Modal
jest.mock('antd', () => {
    const original = jest.requireActual('antd');
    return {
        ...original,
    };
});

describe('Test Function Manage', () => {
    beforeEach(() => {
        cleanup();
        document.body.innerHTML = '';
        (catalogueService.loadTreeNode as jest.Mock).mockReset();
        const mockFolderTree: any = {
            data: [
                {
                    id: 1,
                    name: 'root',
                    data: {
                        level: 1,
                    },
                },
                {
                    id: 2,
                    name: 'folder',
                    data: {
                        id: 2,
                        name: 'test',
                        parentId: 1,
                    },
                },
                {
                    id: 3,
                    name: 'file',
                    isLeaf: true,
                    data: {
                        type: 'file',
                    },
                },
            ],
        };

        const mo = document.createElement('div');
        mo.id = 'molecule';
        document.body.appendChild(mo);

        (FolderTree as unknown as jest.Mock)
            .mockReset()
            .mockImplementation(({ folderTree = mockFolderTree, onRightClick, onSelectFile, onClickContextMenu }) => (
                <div data-testid="mockFolderTree">
                    {folderTree?.data?.map((i: any) => (
                        <div
                            key={i.id}
                            data-testid="mockFolderTreeNode"
                            onContextMenu={() => {
                                const context = onRightClick(i) as {
                                    id: string;
                                    name: string;
                                }[];

                                if (document.querySelector('#mockFolderTreeContextMenu')) {
                                    const div = document.querySelector<HTMLDivElement>('#mockFolderTreeContextMenu')!;

                                    div.innerHTML = '';

                                    context.forEach((c) => {
                                        const node = document.createElement('div');
                                        node.dataset.testid = c.id;
                                        node.innerText = c.name;
                                        node.onclick = () => {
                                            onClickContextMenu(c, i);
                                        };
                                        div.appendChild(node);
                                    });
                                } else {
                                    const div = document.createElement('div');
                                    div.dataset.testid = 'mockFolderTreeContextMenu';
                                    div.id = 'mockFolderTreeContextMenu';

                                    context.forEach((c) => {
                                        const node = document.createElement('div');
                                        node.dataset.testid = c.id;
                                        node.innerText = c.name;
                                        node.onclick = () => {
                                            onClickContextMenu(c, i);
                                        };
                                        div.appendChild(node);
                                    });

                                    document.body.appendChild(div);
                                }
                            }}
                            onClick={() => onSelectFile(i)}
                        >
                            {i.name}
                        </div>
                    ))}
                </div>
            ));
    });

    it('Should match snapshot', () => {
        const { asFragment } = render(
            <Function
                panel={{
                    id: 'FunctionManager',
                    icon: 'variable-group',
                    name: '函数管理',
                    title: '函数管理',
                }}
                headerToolBar={[
                    {
                        id: 'refresh',
                        title: '刷新',
                        icon: 'refresh',
                    },
                ]}
            />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('Should support create new folder', () => {
        const { getAllByTestId, getByTestId } = render(
            <Function
                panel={{
                    id: 'FunctionManager',
                    icon: 'variable-group',
                    name: '函数管理',
                    title: '函数管理',
                }}
                headerToolBar={[
                    {
                        id: 'refresh',
                        title: '刷新',
                        icon: 'refresh',
                    },
                ]}
            />
        );

        const nodes = getAllByTestId('mockFolderTreeNode');

        fireEvent.contextMenu(nodes[0]);
        [FUNCTOIN_ACTIONS.CREATE_FUNCTION, FUNCTOIN_ACTIONS.CREATE_FOLDER].forEach((item) => {
            expect(getByTestId(item.id)).toBeInTheDocument();
        });

        fireEvent.contextMenu(nodes[1]);
        [
            FUNCTOIN_ACTIONS.CREATE_FUNCTION,
            FUNCTOIN_ACTIONS.CREATE_FOLDER,
            FUNCTOIN_ACTIONS.EDIT,
            FUNCTOIN_ACTIONS.DELETE,
        ].forEach((item) => {
            expect(getByTestId(item.id)).toBeInTheDocument();
        });

        fireEvent.contextMenu(nodes[2]);
        [FUNCTOIN_ACTIONS.EDIT, FUNCTOIN_ACTIONS.DELETE].forEach((item) => {
            expect(getByTestId(item.id)).toBeInTheDocument();
        });
    });

    it("Should support view function's detail", async () => {
        (api.getOfflineFn as jest.Mock).mockReset().mockResolvedValue({
            code: 1,
        });

        const { getAllByTestId } = render(
            <Function
                panel={{
                    id: 'FunctionManager',
                    icon: 'variable-group',
                    name: '函数管理',
                    title: '函数管理',
                }}
                headerToolBar={[
                    {
                        id: 'refresh',
                        title: '刷新',
                        icon: 'refresh',
                    },
                ]}
            />
        );

        const nodes = getAllByTestId('mockFolderTreeNode');

        await act(async () => {
            fireEvent.click(nodes[0]);
        });
        expect(functionManagerService.setActive).lastCalledWith(1);
        expect(api.getOfflineFn).not.toBeCalled();

        await act(async () => {
            fireEvent.click(nodes[2]);
        });
        expect(functionManagerService.setActive).lastCalledWith(3);
        expect(api.getOfflineFn).toBeCalled();
    });

    it('Should support create new function', async () => {
        (api.addOfflineFunction as jest.Mock).mockReset().mockResolvedValue({
            code: 1,
        });
        (TreeViewUtil as jest.Mock).mockReset().mockImplementation(() => ({
            getHashMap: jest
                .fn()
                .mockImplementationOnce(() => ({
                    node: {
                        data: {
                            catalogueType: 'FunctionManager',
                        },
                    },
                    parent: {},
                }))
                .mockImplementation(() => ({
                    node: {
                        data: {
                            catalogueType: 'SparkSQLFunction',
                        },
                    },
                })),
        }));

        const { getAllByTestId, getByTestId, getByText } = render(
            <Function
                panel={{
                    id: 'FunctionManager',
                    icon: 'variable-group',
                    name: '函数管理',
                    title: '函数管理',
                }}
                headerToolBar={[
                    {
                        id: 'refresh',
                        title: '刷新',
                        icon: 'refresh',
                    },
                ]}
            />
        );

        const nodes = getAllByTestId('mockFolderTreeNode');

        // Open modal
        fireEvent.contextMenu(nodes[0]);
        fireEvent.click(getByTestId(FUNCTOIN_ACTIONS.CREATE_FUNCTION.id));

        await waitFor(() => {
            expect(getByText('新建自定义函数')).toBeInTheDocument();
        });

        // Click cancel button
        fireEvent.click(document.querySelector('.ant-modal-footer')!.querySelectorAll('button')[0]);

        await waitFor(() => {
            expect(document.querySelector('#test-id')).toBeNull();
        });

        // Open modal again
        fireEvent.click(getByTestId(FUNCTOIN_ACTIONS.CREATE_FUNCTION.id));
        await waitFor(() => {
            expect(getByText('新建自定义函数')).toBeInTheDocument();
        });

        fillFormContent(getAllByTestId);
        fireEvent.click(document.querySelector('.ant-modal-footer')!.querySelectorAll('button')[1]);

        await waitFor(() => {
            expect(api.addOfflineFunction).toBeCalledWith({
                className: 'com.dtstack',
                commandFormate: 'main',
                name: 'mock',
                nodePid: '1',
                paramDesc: 'test',
                purpose: 'forTest',
                resourceId: '1',
                taskType: 5,
                udfType: 0,
            });
            // Refresh parent node after creating
            expect(catalogueService.loadTreeNode).toBeCalled();
        });
    });

    it('Should support create new folder', async () => {
        (api.addOfflineCatalogue as jest.Mock).mockReset().mockResolvedValue({
            code: 1,
        });
        const { getAllByTestId, getByTestId, getByText } = render(
            <Function
                panel={{
                    id: 'FunctionManager',
                    icon: 'variable-group',
                    name: '函数管理',
                    title: '函数管理',
                }}
                headerToolBar={[
                    {
                        id: 'refresh',
                        title: '刷新',
                        icon: 'refresh',
                    },
                ]}
            />
        );

        const nodes = getAllByTestId('mockFolderTreeNode');
        // Open modal
        fireEvent.contextMenu(nodes[0]);
        fireEvent.click(getByTestId(FUNCTOIN_ACTIONS.CREATE_FOLDER.id));

        await waitFor(() => {
            expect(getByText('新建文件夹')).toBeInTheDocument();
        });

        fireEvent.change(document.querySelector('input#dt_nodeName')!, {
            target: { value: 'test' },
        });
        fireEvent.change(getByTestId('mockFolderPicker'), { target: { value: 1 } });

        fireEvent.click(document.querySelector('.ant-modal-footer')!.querySelectorAll('button')[1]);

        await waitFor(() => {
            expect(api.addOfflineCatalogue).toBeCalledWith({
                nodeName: 'test',
                nodePid: '1',
            });
        });
    });

    it('Should support edit function', async () => {
        (api.getOfflineFn as jest.Mock).mockReset().mockResolvedValue({
            code: 1,
            data: {
                id: 1,
                className: 'com.dtstack',
                commandFormate: 'main',
                name: 'mock',
                nodePid: 1,
                paramDesc: 'test',
                purpose: 'forTest',
                resources: 1,
                taskType: 5,
                udfType: 0,
            },
        });
        (api.addOfflineFunction as jest.Mock).mockReset().mockResolvedValue({
            code: 1,
        });
        const { getAllByTestId, getByTestId, getByText } = render(
            <Function
                panel={{
                    id: 'FunctionManager',
                    icon: 'variable-group',
                    name: '函数管理',
                    title: '函数管理',
                }}
                headerToolBar={[
                    {
                        id: 'refresh',
                        title: '刷新',
                        icon: 'refresh',
                    },
                ]}
            />
        );

        const nodes = getAllByTestId('mockFolderTreeNode');
        // Open modal
        fireEvent.contextMenu(nodes[2]);
        fireEvent.click(getByTestId(FUNCTOIN_ACTIONS.EDIT.id));

        await waitFor(() => {
            expect(api.getOfflineFn).toBeCalled();
        });
        await waitFor(() => {
            expect(getByText('编辑自定义函数')).toBeInTheDocument();
        });

        fireEvent.change(document.querySelector('input#nodePid')!, { target: { value: 11 } });
        fireEvent.click(document.querySelector('.ant-modal-footer')!.querySelectorAll('button')[1]);

        await waitFor(() => {
            expect(api.addOfflineFunction).toBeCalledWith({
                className: 'com.dtstack',
                commandFormate: 'main',
                id: 1,
                name: 'mock',
                nodePid: '11',
                paramDesc: 'test',
                purpose: 'forTest',
                resourceId: 1,
                resources: 1,
                taskType: 5,
                udfType: 0,
            });
        });
    });

    it('Should support edit folder', async () => {
        (api.editOfflineCatalogue as jest.Mock).mockReset().mockResolvedValue({
            code: 1,
        });
        const { getAllByTestId, getByTestId, getByText } = render(
            <Function
                panel={{
                    id: 'FunctionManager',
                    icon: 'variable-group',
                    name: '函数管理',
                    title: '函数管理',
                }}
                headerToolBar={[
                    {
                        id: 'refresh',
                        title: '刷新',
                        icon: 'refresh',
                    },
                ]}
            />
        );

        const nodes = getAllByTestId('mockFolderTreeNode');
        // Open modal
        fireEvent.contextMenu(nodes[1]);
        fireEvent.click(getByTestId(FUNCTOIN_ACTIONS.EDIT.id));

        await waitFor(() => {
            expect(getByText('编辑文件夹')).toBeInTheDocument();
        });

        fireEvent.change(document.querySelector('input#dt_nodeName')!, {
            target: { value: 'mock' },
        });
        fireEvent.click(document.querySelector('.ant-modal-footer')!.querySelectorAll('button')[1]);

        await waitFor(() => {
            expect(api.editOfflineCatalogue).toBeCalledWith({
                id: 2,
                nodeName: 'mock',
                nodePid: 1,
                type: 'folder',
            });
        });
    });

    it('Should support delete folder and function', async () => {
        (api.delOfflineFolder as jest.Mock).mockReset().mockResolvedValue({
            code: 1,
        });
        (api.delOfflineFn as jest.Mock).mockReset().mockResolvedValue({
            code: 1,
        });
        const { getAllByTestId, getByTestId, getByText } = render(
            <Function
                panel={{
                    id: 'FunctionManager',
                    icon: 'variable-group',
                    name: '函数管理',
                    title: '函数管理',
                }}
                headerToolBar={[
                    {
                        id: 'refresh',
                        title: '刷新',
                        icon: 'refresh',
                    },
                ]}
            />
        );

        const nodes = getAllByTestId('mockFolderTreeNode');
        // Open modal
        fireEvent.contextMenu(nodes[1]);
        fireEvent.click(getByTestId(FUNCTOIN_ACTIONS.DELETE.id));

        await waitFor(() => {
            expect(getByText('确认要删除此文件夹吗?')).toBeInTheDocument();
        });

        await act(async () => {
            fireEvent.click(document.querySelector('.ant-modal-confirm-btns')!.querySelectorAll('button')[1]);
        });

        await waitFor(() => {
            expect(api.delOfflineFolder).toBeCalled();
        });
    });
});
