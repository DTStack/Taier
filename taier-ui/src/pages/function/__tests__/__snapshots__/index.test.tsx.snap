// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Function Manage Should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="functionManager-container"
  >
    <div
      data-testid="mockHeader"
    >
      函数管理
      <div
        data-testid="mockHeaderToolbar"
      >
        <div
          data-testid="ActionBar"
        >
          <div
            id="refresh"
          />
        </div>
      </div>
    </div>
    <div
      data-testid="mockContent"
    >
      <div
        class="functionManager-content"
        tabindex="0"
      >
        <div
          data-testid="mockFolderTree"
        >
          <div
            data-testid="mockFolderTreeNode"
          >
            root
          </div>
          <div
            data-testid="mockFolderTreeNode"
          >
            folder
          </div>
          <div
            data-testid="mockFolderTreeNode"
          >
            file
          </div>
        </div>
      </div>
    </div>
    <div />
  </div>
</DocumentFragment>
`;
